### ISSUE TITLE
Question: Background and explaining set sizes for TreeExplainer'ing large data sets

### ISSUE THREAD

**Issue by parrt** [shap/shap#955]
From looking through the other issues and the documentation in the source code, I think I'm coming to understand what data needs to go where when using SHAP. I'm using a random forest and therefore the tree explainer. Here is my current understanding:

* `TreeExplainer(model, data=mybackset, feature_dependence="interventional")` requires a "background set" to "break dependencies" between codependent variables, meaning be able to treat variables independently during simulation; cost scales linearly with `len(mybackset)`
* `TreeExplainer(model, feature_dependence="tree_path_dependent")` simply uses the branching of the tree itself to provide explanations without need to directly access the original training data
* Runtime speed for explanations is most heavily affected by the complexity of the RF, and is quadratic in the tree depth

Then we use `shap_values(Xe)`  to get the actual values for records in some `Xe` explanation set. The cost is linear in `len(Xe)`.

For speed reasons, it's clear we need to reduce the size of the background set, when needed to explain codependent variables, and the number of records in the explanation set.  Fair enough. My questions are:

1. How much data is needed in the background set to break dependencies between variables? Does it depend on how many variables are codependent with each other or the sets of variables that are codependent? Your recommendation is to use less than 1000 background set examples but for a data set with, say, 40,000 records, is 1000 big enough to separate variables?
2.  For a large data set, how few records in the explanatory set can we get away with? Given the quadratic complexity on the tree depth, even a few test records seems expensive. For example, using the league of legends data set from your notebook and training a modest 50-tree random forest with 1229705 records, I'm seeing 1.5s per explanatory record. For 1000 records, that would take 25 minutes. Would 1000 records be enough to adequately summarize the effect of variables among 1229705 records?

Even if slow, SHAP provides answers to questions we can't get right now, so I'm not complaining haha, but it's still important to understand how well we are summarizing data given the amount of information provided to SHAP.

Do you have any anecdotal evidence or experiments that suggest how much data is needed for background sets and explanatory sets to capture the underlying relationships in a data set?

**Comment by slundberg** [comment]
These are great questions!

Your description of the time bounds is right, except that when `feature_dependence="interventional"` the runtime is O(TLN) for each test sample you want to explain, where T is the number of trees in the ensemble, L is the number of leaves in each tree, and N is the number of background samples you provide. So it is not quadratic in the depth of the tree for this case.

For `feature_dependence="tree_path_dependent"` it uses the coverage information in the tree and so you get O(TLD^2) runtime where D is the depth of the trees.

For gradient boosted trees things typically run really fast, but for random forests with huge trees things can run much slower as you have seen. Now to your questions:

1. You only need one sample to break dependencies between the features. Using one sample is commonly used in deep learning scenarios where people provide a single "reference input" and explain things relative to that reference. The issue here is that your background samples represent your prior distribution over the input space, and using a point prior is often not what you want. To well represent a population you might think you need to cover the whole high dimensional space, but this is not true since we are exactly computing the SHAP values for each reference and then doing simple averaging over the references. This means we will converge at the standard `sqrt(N)` rate of mean convergence. So the standard deviation of your estimate with 1000 background samples is 10 times smaller than with 10 samples. It is important to note though that you don't really need to converge here to get a good explanation. Often people have a vauge prior in mind anyway, and so anything that roughly represents the background will work well (which is why some other methods recommend using single reference values).

2. This is a much more quantifiable question, since you are now asking how many examples do I need to explain to see patterns in my data. This of course depends on the pattern and its complexity. The best advice I have is to run a bootstrap test and see the variability. Of course if you are trying to save time this is not the way to go, but you could do something similar like run 3 sets of 100 and see if you see a consistent patten in each of the three groups before merging them. Again the sqrt(N) convergence shows up for any pattern that is an average (like the global effect of a feature). I explain 1k-10k samples typically for visualizations. And I would run a full end-to-end bootstrap analysis (including model training) if I want to know if a pattern is consistent in the data (and not just the particular model I have trained). Note also that less than 100 samples could be plenty to detect important changes in distributions during model monitoring where we are explaining the loss of the model and watching the input feature effects over time.

Not sure if that is helpful, but it at least tells you what I know :)

**Comment by keyur-potdar** [comment]
@slundberg that's a great explanation. 

What's the minimum number of background set samples we can use when we have too many classes? For example, if I have a dataset having 1000 classes, will 100 samples be enough as the background set to give good explanations?

**Comment by hipoglucido** [comment]
Hey, I am using the interventional approach with background dataset for my binary classification imbalanced problem with sklearn's RF. My training dataset is created by downsampling the majority class. My question is, should the background dataset have the downsampling applied as well? Of course, the prediction dataset used in PR won't have any downsampling applied since I don't know the labels
